### PR TITLE
TF-M: regular upmerge, post Zephyr v2.6.0

### PR DESCRIPTION
**Pull request by ioannisg**
- TF-M upmerge with several fixes. 
  upmerge up to upstream commit SHA: a199c3047f320a2f82b9a0c27af5b50991184e0f
  In addition to the fixes, the upmerge brings in the new FLIH service.
- We remove the TF-M specific module for mcuboot. TF-M builds in Zephyr will now use the mcuboot that Zephyr comes with.

### CODE REVIEW

**Comment by microbuilder** [comment]
Regression tests are failing for PS in QEMU, but this seems to be an issue with either QEMU or the upstream TF-M project, not with Zephyr.

The issue can be reproduce using only upstream TF-M by cloning TF-M and running the following script. It has been tested on QEMU 6.0.0 (by myself) andd 5.1.0 (Ioannis):

```bash
#!/bin/bash

# Exit on error
set -e

# Run from 'trusted-firmware-m'
rm -rf cmake_build
cmake -S . -B cmake_build -DTFM_PLATFORM=arm/mps2/an521 -DTFM_TOOLCHAIN_FILE=toolchain_GNUARM.cmake -DCMAKE_BUILD_TYPE=Debug -DTEST_S=ON -DTEST_NS=ON
cmake --build cmake_build -- install

# Generate a merged binary
srec_cat cmake_build/install/outputs/ARM/MPS2/AN521/tfm_s_ns_signed.bin -Binary -offset=0x80000 -o tfm_s_ns_signed.hex -Intel
srec_cat cmake_build/install/outputs/ARM/MPS2/AN521/bl2.hex -Intel tfm_s_ns_signed.hex -Intel -o tfm_full.hex -Intel

# Optional: Run in qemu
qemu-system-arm -M mps2-an521 -device loader,file=tfm_full.hex -serial stdio
```

This results in the regression tests consistently hanging in the PS test suite:

```
*** End of Secure test suites ***

#### Execute test suites for the Non-secure area ####
Running Test Suite PSA protected storage NS interface tests (TFM_NS_PS_TEST_1XXX)...
> Executing 'TFM_NS_PS_TEST_1001' 
  Description: 'Set interface'
  TEST: TFM_NS_PS_TEST_1001 - PASSED!
> Executing 'TFM_NS_PS_TEST_1002' 
  Description: 'Set interface with create flags'
  TEST: TFM_NS_PS_TEST_1002 - PASSED!
> Executing 'TFM_NS_PS_TEST_1003' 
  Description: 'Set interface with NULL data pointer'
  TEST: TFM_NS_PS_TEST_1003 - PASSED!
> Executing 'TFM_NS_PS_TEST_1004' 
  Description: 'Set interface with write once UID'
  TEST: TFM_NS_PS_TEST_1004 - PASSED!
> Executing 'TFM_NS_PS_TEST_1005' 
  Description: 'Get interface with valid data'
  TEST: TFM_NS_PS_TEST_1005 - PASSED!
> Executing 'TFM_NS_PS_TEST_1006' 
  Description: 'Get interface with zero data length'
  TEST: TFM_NS_PS_TEST_1006 - PASSED!
> Executing 'TFM_NS_PS_TEST_1007' 
  Description: 'Get interface with invalid UIDs'
  TEST: TFM_NS_PS_TEST_1007 - PASSED!
> Executing 'TFM_NS_PS_TEST_1008' 
  Description: 'Get interface with invalid data lengths and offsets'
  TEST: TFM_NS_PS_TEST_1008 - PASSED!
> Executing 'TFM_NS_PS_TEST_1009' 
  Description: 'Get interface with NULL data pointer'
  TEST: TFM_NS_PS_TEST_1009 - PASSED!
> Executing 'TFM_NS_PS_TEST_1010' 
  Description: 'Get info interface with write once UID'
  TEST: TFM_NS_PS_TEST_1010 - PASSED!
> Executing 'TFM_NS_PS_TEST_1011' 
  Description: 'Get info interface with valid UID'
  TEST: TFM_NS_PS_TEST_1011 - PASSED!
> Executing 'TFM_NS_PS_TEST_1012' 
  Description: 'Get info interface with invalid UIDs'
  TEST: TFM_NS_PS_TEST_1012 - PASSED!
> Executing 'TFM_NS_PS_TEST_1013' 
  Description: 'Remove interface with valid UID'
```

**Comment by ioannisg** [comment]
@microbuilder I will disable the test temporarily for mps2_an521_ns and file issues in Zephyr and TFM upstream.

**Comment by ioannisg** [comment]
> @microbuilder I will disable the test temporarily for mps2_an521_ns and file issues in Zephyr and TFM upstream.

https://developer.trustedfirmware.org/T951

**Review by microbuilder — APPROVED** [review]

**Review by microbuilder — APPROVED** [review]
LGTM

**Review by galak — APPROVED** [review]

**Comment by microbuilder** [comment]
Adding DNM until bootloader issues can be resolved.

**Comment by ioannisg** [comment]
> Adding DNM until bootloader issues can be resolved.

@microbuilder I believe the issues here are unrelated to TF-M. I am not 100% sure why CI needs to run tests on platforms that do not use  TF-M...

**Comment by microbuilder** [comment]
> > Adding DNM until bootloader issues can be resolved.
> 
> @microbuilder I believe the issues here are unrelated to TF-M. I am not 100% sure why CI needs to run tests on platforms that do not use TF-M...

@ioannisg The problem here is with MCUBoot, and the last minute update of MCUBoot in this PR, since there are now two identical references to the same library:

- https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L113
- https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L147

This leads to tthe `MCUBOOT_BOOTUTIL` library being added twice: https://github.com/zephyrproject-rtos/mcuboot/blob/main/boot/bootutil/zephyr/CMakeLists.txt#L7

**Comment by ioannisg** [comment]
> > > Adding DNM until bootloader issues can be resolved.
> > 
> > 
> > @microbuilder I believe the issues here are unrelated to TF-M. I am not 100% sure why CI needs to run tests on platforms that do not use TF-M...
> 
> @ioannisg The problem here is with MCUBoot, and the last minute update of MCUBoot in this PR, since there are now two identical references to the same library:
> 
>     * https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L113
> 
>     * https://github.com/zephyrproject-rtos/zephyr/blob/c078fa45b32e8cc4ea4d2264a1fd2e17a0ac6585/west.yml#L147
> 
> 
> This leads to tthe `MCUBOOT_BOOTUTIL` library being added twice: https://github.com/zephyrproject-rtos/mcuboot/blob/main/boot/bootutil/zephyr/CMakeLists.txt#L7

Yes, I figured this out as well, you're correct. 
to elaborate a bit more:
- some recent MCUboot changes have made MCUBOOT_BOOTUTIL an interface library.
- when we upgrate tfm-mcuboot, we are bringing in this change
- as you point out, this library is added twice in the build, under some configurations

What are our options?
1. Do not allow the build system (cmake) to add the tfm-mcuboot subdirectory, for the current build. This makes sense because tfm-mcuboot is _not_ a Zephyr module; it is used only by the TFM build. We should have the possibility to exclude some "modules" in west from being added to the build. To the best of my knowledge, this is not supported today, but IMHO it should be supported, because we might want to exlcude some module paths from our builds in certain situations (such as this one)
2. Do not use a specific MCUboot version for TF-M in  Zephyr. This is the "right thing to do", in accordance with how we addressed the mbedTLS case: Zephyr and TF-M should have the same versions (if possible) for dependencies such as MCUboot or mbedtls (possibly others).

For the purpose of the current upmerge, I am going to try to go for option (2). But I will also file an issue about (1) for @tejlmand to consider, when he is back.

File: west.yml [hunk 1]
```diff
@@ -162,7 +162,7 @@ manifest:
       groups:
         - crypto
     - name: mcuboot
-      revision: 7a5196820ba48a6ad7e7d573c9048c021960677c
+      revision: pull/55/head
```
Comment (nvlsianpu) [inline west.yml:165]: you can update the SHA

Comment (ioannisg) [reply west.yml]: done now, thanks!

**Pull request merged**